helix-core/src/main/java/org/apache/helix/common/caches/CurrentStateSnapshot.java

```diff
@@ -32,18 +37,32 @@ public CurrentStateSnapshot(final Map<PropertyKey, CurrentState> currentStateMap
     if (_updatedStateKeys != null && _prevStateMap != null) {
       // Note if the prev state map is empty, this is the first time refresh.
       // So the update is not considered as "recent" change.
+      int driftCnt = 0; // clock drift count for comparing timestamp
       for (PropertyKey propertyKey : _updatedStateKeys) {
         CurrentState prevState = _prevStateMap.get(propertyKey);
         CurrentState curState = _properties.get(propertyKey);
 
         Map<String, Long> partitionUpdateEndTimes = null;
         for (String partition : curState.getPartitionStateMap().keySet()) {
           long newEndTime = curState.getEndTime(partition);
-          if (prevState == null || prevState.getEndTime(partition) < newEndTime) {
+          if (prevState == null
+              || prevState.getEndTime(partition) < newEndTime && prevState.getEndTime(partition) != -1) {
             if (partitionUpdateEndTimes == null) {
               partitionUpdateEndTimes = new HashMap<>();
             }
             partitionUpdateEndTimes.put(partition, newEndTime);
+          } else if (prevState != null && prevState.getEndTime(partition) > newEndTime) {
+            // This can happen due to clock drift.
+            // updatedStateKeys is the path to resource in an instance config.
+            // Thus, the space of inner loop is Sigma{replica(i) * partition(i)}; i over all resources in the cluster
+            // This space can be large. In order not to print two many lines, we print first warning for the first case.
+            // If clock drift turns out to be common, we can consider print out more logs, or expose an metric.
+            if (driftCnt < 1) {
```

Comment:
Several comments here.
1. driftCnt as a boolean will be enough?
2. The drift means if 2 participants have a different clock, right? In this case, I strongly prefer to write a separate tool to measure instead of the piggypack that takes advantage of helix participant.
Any warning message might be asked by our customer. Since it might look very scary to some customers, if we add this, we'd better be prepared to answer what's this message about and why we need this.
Moreover, this will just sample the time drift situation. We will not understand to overall picture.
3. In a distributed system, a certain extent of time drift is expected. So ideally, we should be able to tolerate. If we do find this warning in our log, what shall we do? If there is no next step, this log won't provide too much value.

Reply:
1/ boolean is enough. But what if later you decide to log once for every 100 such event. In this case, we can easily change to if (driftCnt % 100 == 0) 

2/ 3/ Currently, the production has NTP (network time protocol) deamon running, which would sync the clock to around 10ms in the same data center. This is something fine for us. 

I think your concern about  "scary" to customer is valid. However, on the other hand, if we don't log this warning, what if there are some "huge spikes" in statePropagationDelay metrics, due to clock drift say NTP config messed up.  We still need to troubleshoot, but this time, without any clue. 

In fact, this time, the database guys actually asked for "huge spikes" statePropagationDelay which is due to the "-1" issue fixed here. 

So weighing the pro and cons from both side. It seems that adding this log is still a better "evil"?

Let me know what is your take here.

Reply:
Sorry, I'm still not convinced. My previous question is that if we debug with this log, and we confirmed there is clock drift, what can we do? Most possibly fixing NTP, right? If that's the case, why not monitoring the NTP status and have some alert on that. This seems more direct and efficient.

Moreover, with this check, what if the drift is to the other direction? We will just take the result and record, right? I just don't think we can do it good enough. This concern (if it really exists) should be addressed as a separate initiative.